### PR TITLE
feat(sqlite): support serialize/deserialize sqlite APis

### PR DESCRIPTION
**Pull request by yuvalp-k2view**

### CODE REVIEW

**Comment by yuvalp-k2view** [comment]
I see some warning about using DirectByteBuffer and recommendations to use byte[]. Reading up on it ... a fix is coming

File: src/main/java/org/sqlite/core/NativeDB.java
Comment (gotson) [inline src/main/java/org/sqlite/core/NativeDB.java:569]: why is that a problem?

File: src/test/java/org/sqlite/SerializeTest.java
Comment (gotson) [inline src/test/java/org/sqlite/SerializeTest.java:48]: probably better to use `assertThatThrownBy`

File: src/test/java/org/sqlite/SerializeTest.java
Comment (gotson) [inline src/test/java/org/sqlite/SerializeTest.java:82]: what does that test ? there's no assertions

File: src/test/java/org/sqlite/SerializeTest.java
Comment (gotson) [inline src/test/java/org/sqlite/SerializeTest.java:99]: should use `assertThatThrownBy` for clarity

**Review by gotson — CHANGES_REQUESTED** [review]

File: src/main/java/org/sqlite/core/NativeDB.java
Comment (yuvalp-k2view) [reply src/main/java/org/sqlite/core/NativeDB.java:569]: ByteBuffer is limited to max int

Comment (gotson) [reply src/main/java/org/sqlite/core/NativeDB.java:569]: can we use something else in that case ? that seems very small

Comment (yuvalp-k2view) [reply src/main/java/org/sqlite/core/NativeDB.java:569]: It's all meant for in memory databases so not that small.
We could use an array of buffers. That would not be a very elegant API though.
I could wrap it with an InputStream so the end user will access it with a more standard approach. Could support both APIs. buffer for small databases and InputStream for larger?
wdyt?

Comment (gotson) [reply src/main/java/org/sqlite/core/NativeDB.java:569]: > It's all meant for in memory databases so not that small.

Not sure to understand this statement. This can be used to serialize any database, not just in-memory, no? Or did i misunderstood something?

>  We could use an array of buffers. That would not be a very elegant API though.

do we have to use a Buffer ? Why not just a byte array?

Comment (yuvalp-k2view) [reply src/main/java/org/sqlite/core/NativeDB.java:569]: Theoretically you could Serialize a file into memory but that would be very inefficient. Makes much more sense to just read the file. Deserialize is by definition memory only.
The max sqlite default for this mode is only 1,073,741,824 (1gb).
I am checking if I can increase it.
There are good reasons to switch to byte[] (might be faster oveall, see comment below). However, it is also limited to max int size.

Comment (yuvalp-k2view) [reply src/main/java/org/sqlite/core/NativeDB.java:569]: More info here: I've managed to raise the limit but anything beyond 2gb results in a malloc error from sqlite.
Turns out it does want to allocate more than 2gb of continuous space. I will ask the sqlite folks if there is a way around it but for now I think we should leave it at that.

File: src/test/java/org/sqlite/SerializeTest.java
Comment (yuvalp-k2view) [reply src/test/java/org/sqlite/SerializeTest.java]: done

File: src/test/java/org/sqlite/SerializeTest.java
Comment (yuvalp-k2view) [reply src/test/java/org/sqlite/SerializeTest.java:63]: Added assert that size doesn't change across runs

File: src/test/java/org/sqlite/SerializeTest.java
Comment (yuvalp-k2view) [reply src/test/java/org/sqlite/SerializeTest.java]: Done

**Comment by gotson** [comment]
it seems there are some CI errors, can you have a look?

File: src/main/java/org/sqlite/core/NativeDB.java
Comment (yuvalp-k2view) [reply src/main/java/org/sqlite/core/NativeDB.java:569]: Asked and answered by sqlite maintainer. 2gb is a hard coded limit in the library, no way to change it, not even with a compilation flag.

**Comment by yuvalp-k2view** [comment]
> it seems there are some CI errors, can you have a look?

ran spotless:apply o that should be okay.

The unsatisfied link error ... I'm baffled by. I was able to link and run all the tests locally without a problem and of course the native code contains the Java_org_sqlite_core_NativeDB_serializeSize function.

Any ideas? Anything I can do locally to reproduce?

**Comment by yuvalp-k2view** [comment]
It seems the "make" artifacts are committed... didn't occur to me. Running a clean 'make' and will commit them.

**Comment by yuvalp-k2view** [comment]
Binaries comitted

File: src/main/java/org/sqlite/core/NativeDB.java
Comment (gotson) [reply src/main/java/org/sqlite/core/NativeDB.java:569]: might be worth to leave a comment in the code about it

**Comment by gotson** [comment]
yes, binaries are committed to the repo, and can be built [using a PR comment](https://github.com/xerial/sqlite-jdbc/blob/master/CONTRIBUTING.md#build-from-ci) too.

**Comment by gotson** [comment]
@pyckle may i ask whether you would have a bit of time to glance over the C code of that PR, and provide your insights ?

File: src/main/java/org/sqlite/core/NativeDB.java
Comment (yuvalp-k2view) [reply src/main/java/org/sqlite/core/NativeDB.java:569]: Done. Also added the limit to the javadoc comment.

**Comment by yuvalp-k2view** [comment]
@pyckle @gotson 
I see failures only on the graalvm tests. Is there anything I should do to create artifacts or test this?

File: src/main/java/org/sqlite/core/NativeDB.java
Comment (pyckle) [reply src/main/java/org/sqlite/core/NativeDB.java:569]: I believe this limit is slightly wrong.
See: https://stackoverflow.com/questions/3038392/do-java-arrays-have-a-maximum-size
See: https://github.com/OpenHFT/Chronicle-Bytes/blob/a02a79ed8ce9900f8748ad6f3ce801fa241844d1/src/main/java/net/openhft/chronicle/bytes/Bytes.java#L110

I would go with the more conservative limit that openhft suggests. The real limit of a byte array size is JVM dependent. In addition, I don't think this is reachable anyways due to the sqlite malloc limit.

**Comment by pyckle** [comment]
> @pyckle may i ask whether you would have a bit of time to glance over the C code of that PR, and provide your insights ?

I took a quick look. Looks good to me.

> I see failures only on the graalvm tests.

Most likely this is graal native image failures because https://github.com/xerial/sqlite-jdbc/blob/master/src/main/java9/org/sqlite/nativeimage/SqliteJdbcFeature.java needs to be updated to list the newly added JNI calls

File: src/main/java/org/sqlite/core/NativeDB.java
Comment (pyckle) [inline src/main/java/org/sqlite/core/NativeDB.java:573]: It seems to me that it would be better to allocate this array in JNI and return it, and call serialize once.
Even though this DB can't be manipulated here due to the synchronize, there's nothing stopping another connection (or even another process) from manipulating the DB between these calls.

I don't think this is critical, as the JNI code in serialize double checks this and as you said, this is really more for inmemory databases. But it seems minor to fix and would make this more robust.

File: src/main/java/org/sqlite/core/NativeDB.java
Comment (yuvalp-k2view) [reply src/main/java/org/sqlite/core/NativeDB.java:569]: @pyckle I think every module should take care of its own limitations. It is true that the limitation on the sqlite is a bit more lower than the java side so this shouldn't happen (added a comment to that effect).

I did develop a unit test that shows the database can grow up to ~2gb. It of course needs at least 4gb of free memory and takes around 30 seconds. Do you think I should commit it? I thought it was a bit extreme and could fail on some test environments.

**Comment by yuvalp-k2view** [comment]
@pyckle I looked at https://github.com/xerial/sqlite-jdbc/blob/master/src/main/java9/org/sqlite/nativeimage/SqliteJdbcFeature.java
It seems to only list callback methods and not jni calls.
I did add byte[] line as it seems that any type of class to make the bridge is listed there.
What am I missing? Worth a rerun?

**Comment by gotson** [comment]
> @pyckle I looked at [`master`/src/main/java9/org/sqlite/nativeimage/SqliteJdbcFeature.java](https://github.com/xerial/sqlite-jdbc/blob/master/src/main/java9/org/sqlite/nativeimage/SqliteJdbcFeature.java?rgh-link-date=2023-11-02T13%3A48%3A17Z) It seems to only list callback methods and not jni calls. I did add byte[] line as it seems that any type of class to make the bridge is listed there. What am I missing? Worth a rerun?

@kkriske would you be able to have a quick look at this PR, see if we need to adjust anything for the GraalVM native configuration?

File: src/main/java/org/sqlite/core/NativeDB.java
Comment (gotson) [reply src/main/java/org/sqlite/core/NativeDB.java:569]: > I did develop a unit test that shows the database can grow up to ~2gb. It of course needs at least 4gb of free memory and takes around 30 seconds. Do you think I should commit it? I thought it was a bit extreme and could fail on some test environments.

the github workers only have 7gb of ram and 15gb of disk space (total, not available)

File: src/main/java/org/sqlite/core/NativeDB.java
Comment (pyckle) [reply src/main/java/org/sqlite/core/NativeDB.java:573]: > I found no way to this other than allocating DirectByteBuffer in jni
Why not use `NewByteArray()`? It should yield the same performance as allocating the byte array in java. See: `https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/functions.html#New_PrimitiveType_Array_routines`
You can change the native function to return `byte[]`.

For InMemory DBs, concurrent modification obviously isn't an issue, but this may eventually be used for file dbs.

File: src/main/java/org/sqlite/core/NativeDB.java
Comment (pyckle) [reply src/main/java/org/sqlite/core/NativeDB.java:569]: This is up to @gotson as he's the maintainer, but I would suggest committing it with an `@Ignore` annotation. That way we won't have stability issues on CI and if issues with large DB size arise in the future, one can run this test locally.

**Comment by pyckle** [comment]
> @pyckle I looked at https://github.com/xerial/sqlite-jdbc/blob/master/src/main/java9/org/sqlite/nativeimage/SqliteJdbcFeature.java It seems to only list callback methods and not jni calls. I did add byte[] line as it seems that any type of class to make the bridge is listed there. What am I missing? Worth a rerun?

I don't have time to look at this in detail hands on until later next week. If it's still broken then, I'll investigate.

File: src/main/java/org/sqlite/core/NativeDB.java
Comment (yuvalp-k2view) [reply src/main/java/org/sqlite/core/NativeDB.java:569]: Done as you suggested, testVeryLarge is committed and Disabled

File: src/main/java9/org/sqlite/nativeimage/SqliteJdbcFeature.java
Comment (kkriske) [inline src/main/java9/org/sqlite/nativeimage/SqliteJdbcFeature.java:127]:
```suggestion
```

File: src/test/java/org/sqlite/SerializeTest.java
Comment (kkriske) [inline src/test/java/org/sqlite/SerializeTest.java:10]:
```suggestion
import java.sql.*;
import org.assertj.core.api.InstanceOfAssertFactories;
import org.junit.jupiter.api.Disabled;
```

File: src/test/java/org/sqlite/SerializeTest.java
Comment (kkriske) [inline src/test/java/org/sqlite/SerializeTest.java:86]:
```suggestion
                .asInstanceOf(InstanceOfAssertFactories.type(SQLException.class))
                .extracting(SQLException::getErrorCode)
                .isEqualTo(SQLiteErrorCode.SQLITE_NOTADB.code);
```

File: src/test/java/org/sqlite/SerializeTest.java
Comment (kkriske) [inline src/test/java/org/sqlite/SerializeTest.java:117]:
```suggestion
                    .asInstanceOf(InstanceOfAssertFactories.type(SQLiteException.class))
                    .extracting(SQLiteException::getResultCode)
                    .isEqualTo(SQLiteErrorCode.SQLITE_ERROR);
```

File: src/test/java/org/sqlite/SerializeTest.java
Comment (kkriske) [inline src/test/java/org/sqlite/SerializeTest.java:197]:
```suggestion
                    .asInstanceOf(InstanceOfAssertFactories.type(SQLException.class))
                    .extracting(SQLException::getErrorCode)
                    .isEqualTo(SQLiteErrorCode.SQLITE_FULL.code);
```

**Review by kkriske — CHANGES_REQUESTED** [review]
I don't have write rights so couldn't add this as a commit. I've added suggestions where needed.

The problem is purely in tests, `hasFieldOrPropertyWithValue` uses reflective lookups to the relevant field, so this would require native-image configuration. Either way, using such reflection is bad form imo as you're testing hidden internals and not exposed API. The suggestions remove the `hasFieldOrPropertyWithValue` calls and replace them with proper casts and method calls.

File: src/main/java9/org/sqlite/nativeimage/SqliteJdbcFeature.java
Comment (yuvalp-k2view) [reply src/main/java9/org/sqlite/nativeimage/SqliteJdbcFeature.java]: removed

File: src/test/java/org/sqlite/SerializeTest.java
Comment (yuvalp-k2view) [reply src/test/java/org/sqlite/SerializeTest.java]: did it a bit different

File: src/test/java/org/sqlite/SerializeTest.java
Comment (yuvalp-k2view) [reply src/test/java/org/sqlite/SerializeTest.java]: did it a bit different

File: src/main/java/org/sqlite/core/NativeDB.java
Comment (yuvalp-k2view) [reply src/main/java/org/sqlite/core/NativeDB.java]: Done as you suggested, wasn't aware. Benchmarked, all is well. I also removed the serializedSize method. Don't need it + worried that someone will use it to find out the db size which is very wasteful compared to querying sqlite.

File: src/test/java/org/sqlite/SerializeTest.java
Comment (kkriske) [reply src/test/java/org/sqlite/SerializeTest.java]: That's fine, one minor note: All the `isInstanceOf` checks should probably check against `SQLiteException` and not `SQLException` to be in line with the casts to `SQLiteException`.

**Comment by yuvalp-k2view** [comment]
I've removed reflection from the tests. We can try to run again.

File: src/test/java/org/sqlite/SerializeTest.java
Comment (yuvalp-k2view) [reply src/test/java/org/sqlite/SerializeTest.java]: I started of testing against the contract, but you are right, if I cast, I should check it correctly. done.

**Comment by gotson** [comment]
@yuvalp-k2view could you rebase on the latest `master` ? I have updated sqlite to 3.44.0.

Once done we can rebuild the native libs [with a PR comment](https://github.com/xerial/sqlite-jdbc/blob/master/CONTRIBUTING.md#build-from-ci)

**Comment by yuvalp-k2view** [comment]
I rebased but did not rebuild. I am assuming you will override the binaries

**Comment by gotson** [comment]
it seems the bot can't push to your remote https://github.com/xerial/sqlite-jdbc/actions/runs/6768202925/job/18392278956?pr=1005

**Comment by yuvalp-k2view** [comment]
I rebuilt and pushed

**Comment by yuvalp-k2view** [comment]
Hi @gotson, are there any outstanding issues? Would love to see this feature as part of the 3.44 release.

 (3.44 removes a global lock when creating transaction for in-memory databases. combination of this and serialize is a very serious performance boost).

File: src/main/java/org/sqlite/core/NativeDB.java
Comment (pyckle) [reply src/main/java/org/sqlite/core/NativeDB.java]: Looks good to me. Nice work (:

**Review by gotson — APPROVED** [review]
LGTM

**Pull request merged**
